### PR TITLE
retry eyes test failures only when non-eyes errors occur

### PR DESCRIPTION
**Pull request by davidsbailey**
### Background

the following call tree illustrates roughly how test flakiness is computed:

* dashboard/test/ui/runner.rb
  * cucumber
    * eyes_steps.rb → @eyes.close(throw_exception = true)
    * connect.rb → post success status to saucelabs rest api
* bin/test_flakiness → get success status from saucelabs rest api

"magic retry" uses test flakiness data to decide how many times to rerun each UI test feature. the goal is to enable this for eyes tests during DTTs (but not on Circle), only counting selenium errors (not eyes diffs) as flaky failures.

### Description

1. in eyes_steps.rb
  (a) make eyes mismatches no longer cause the cucumber run to fail, causing these runs to be reported as successful in terms of test flakiness (described briefly above).
  (b) print eyes mismatch errors to the cucumber log
2. in  runner.rb
  (a) parse the cucumber output to detect eyes mismatches
  (b) fail the test run if eyes mismatch is detected
3. in test.rake, turn on "magic retry" for eyes tests

### Screenshots

The following looks different when the only eyes failures occur:

in slack (note "1 eyes mismatches")
screen shot 2017-12-14 at 12 13 12 am

on the command line  (note "1 eyes mismatches"):
screen shot 2017-12-13 at 8 36 39 pm

in the cucumber logs (note red warning text and linkified eyes link):
screen shot 2017-12-13 at 11 54 16 pm

### Testing

Manually verified locally and on the test machine:
* features with only selenium errors show no change in behavior
* passing features show no change in behavior
* features with selenium errors AND eyes mismatches appear just like features with only selenium errors

### CODE REVIEW

**Comment by islemaster** [comment]
:eye_speech_bubble: Reading now....

File: dashboard/test/ui/features/step_definitions/eyes_steps.rb [hunk 1]
```diff
@@ -61,3 +69,12 @@ def ensure_eyes_available
...
   @eyes.host_os = ENV['APPLITOOLS_HOST_OS']
 end
+
+# Wraps any http or https url text in an <a> tag pointing to the same url.
+# Only needs to work on urls like:
+# https://eyes.applitools.com/app/batches/00112233/44556677?accountId=abcdefg.
+# Note that applitools strips the trailing dot on this url for us, so we
+# don't need to strip it here.
+def linkify(str)
+  str.to_s.gsub(/(https?:\/\/[^ ]+)/, '<a href="\1">\1</a>')
```
Comment (islemaster) [inline dashboard/test/ui/features/step_definitions/eyes_steps.rb]: I'd replace `[^ ]` with `\S` so any whitespace ends the link.  Alternatively, there are a number of gems for this sort of thing.

File: dashboard/test/ui/features/step_definitions/eyes_steps.rb [hunk 2]
```diff
@@ -44,7 +48,11 @@
 
   @browser = @original_browser
   fail_on_mismatch = !CDO.ignore_eyes_mismatches
-  @eyes.close(fail_on_mismatch)
+  begin
+    @eyes.close(fail_on_mismatch)
+  rescue Applitools::TestFailedError => e
+    puts "<span style=\"color: red;\">#{EYES_ERROR_PREFIX} #{linkify(e)}</span>"
```
Comment (islemaster) [inline dashboard/test/ui/features/step_definitions/eyes_steps.rb]: Which log does this puts go to?

File: dashboard/test/ui/features/step_definitions/eyes_steps.rb [hunk 3]
```diff
@@ -8,6 +8,10 @@
 # See http://support.applitools.com/customer/en/portal/articles/2099488-match-timeout
 MATCH_TIMEOUT = 5
 
+# Prefix to print before any eyes errors. We depend on this in runner.rb to detect
+# whether there were any eyes errors.
+EYES_ERROR_PREFIX = '[Applitools Eyes error]'
```
Comment (islemaster) [inline dashboard/test/ui/features/step_definitions/eyes_steps.rb]: Could we load this constant from a shared location in both eyes_steps.rb and runner.rb?

File: dashboard/test/ui/features/step_definitions/eyes_steps.rb (hunk 2 above)
Comment (davidsbailey) [reply dashboard/test/ui/features/step_definitions/eyes_steps.rb]: This html ends up in the cucumber log output, e.g. including:
1. the html output (e.g. ChromeLatestWin7_angleHelper_eyes_output.html), and
2. the stdout stream (which is consumed by runner.rb, and may also be what you'd end up with in `success.log` if you didn't specify `runner.rb --html`).

File: dashboard/test/ui/features/step_definitions/eyes_steps.rb (hunk 3 above)
Comment (davidsbailey) [reply dashboard/test/ui/features/step_definitions/eyes_steps.rb]: but then I'd have to pick a name for the file this gets extracted into :-) PTAL

**Pull request merged**